File: problems/141/solution.go
package main
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func hasCycle(head *ListNode) bool {
    if head == nil || head.Next == nil {
        return false
    }
    
    slow, fast := head, head.Next
    
    for fast.Next != nil && fast.Next.Next != nil && fast != slow {
        slow = slow.Next
        fast = fast.Next.Next
    }
    
    return slow == fast
}

File: problems/141/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if not head or not head.next:
            return False
        
        slow, fast = head, head.next
        
        while fast.next and fast.next.next and fast != slow:
            slow = slow.next
            fast = fast.next.next
        
        return slow == fast